administrator/components/com_joomlaupdate/extract.php

```diff
@@ -0,0 +1,1834 @@
+	 * serve 100 requests to extract.php per second and an easy to attack password requiring going over just 1% of
+	 * the search space it'd still take over 282 million seconds to brute force it. Our limit is more than 4 orders
+	 * of magnitude lower than this best practical case scenario, giving us adequate protection against all but the
+	 * luckiest attacker (spoiler alert: the mathematics of probabilities say you're not gonna get too lucky).
+	 *
+	 * It is still advisable to remove the update.php file once you are done with the extraction. This check
+	 * here is only meant as a failsafe in case of a server error during the extraction and subsequent lack of user
+	 * action to remove the update.php file from their server.
+	 */
+	clearstatcache(true);
+	$setupFileCreationTime = filectime($setupFile);
+
+	if (abs(time() - $setupFileCreationTime) > 5400)
+	{
+		return null;
+	}
+
+	// Load update.php. It pulls a variable named $restoration_setup into the local scope.
+	clearFileInOPCache($setupFile);
+
+	require_once $setupFile;
+
+	/** @var array $extractionSetup */
+
+	// The file exists but no configuration is present?
+	if (empty($extractionSetup ?? null) || !is_array($extractionSetup))
+	{
+		return null;
+	}
+
+	/**
+	 * Immediately reject any attempt to run extract.php without a password.
+	 *
+	 * Doing that is a GRAVE SECURITY RISK. It makes it trivial to hack a site. Therefore we are preventing this script
+	 * to run without a password.
+	 */
+	$password     = $extractionSetup['security.password'] ?? null;
+	$userPassword = $_REQUEST['password'] ?? '';
+	$userPassword = !is_string($userPassword) ? '' : trim($userPassword);
+
+	if (empty($password) || !is_string($password) || (trim($password) == '') || (strlen(trim($password)) < 32))
+	{
+		return null;
+	}
+
+	// Timing-safe password comparison. See http://blog.ircmaxell.com/2014/11/its-all-about-time.html
+	if (!timingSafeEquals($password, $userPassword))
+	{
+		return null;
+	}
+
+	// An "instance" variable will resume the engine from the serialised instance
+	$serialized = $_REQUEST['instance'] ?? null;
+
+	if (!is_null($serialized) && empty(ZIPExtraction::unserialiseInstance($serialized)))
+	{
+		// The serialised instance is corrupt or someone tries to trick us. YOU SHALL NOT PASS!
+		return null;
+	}
+
+	return $extractionSetup;
+}
+
+// Import configuration
+$retArray = [
+	'status'  => true,
+	'message' => null,
+];
+
+$configuration = getConfiguration();
+$enabled       = !empty($configuration);
+
+if ($enabled)
+{
+	$sourcePath = $configuration['setup.sourcepath'] ?? '';
+	$sourceFile = $configuration['setup.sourcefile'] ?? '';
+	$destDir    = ($configuration['setup.destdir'] ?? null) ?: __DIR__;
+	$basePath   = rtrim(str_replace('\\', '/', __DIR__), '/');
+	$basePath   = empty($basePath) ? $basePath : ($basePath . '/');
+	$sourceFile = (empty($sourceFile) ? '' : (rtrim($sourcePath, '/\\') . '/')) . $sourceFile;
```

Comment:
```suggestion
	$sourceFile = (empty($sourcePath) ? '' : (rtrim($sourcePath, '/\\') . '/')) . $sourceFile;
```
Shouldn't it be like this?